Update TDF spec for OpenID Connect/Keycloak architecture changes. (#26)
virtru/tdf3-spec: protocol/README.md

## Architecture

The canonical architecture contains four major components.

* *TDF Client* - Initiates and drives the TDF encryption and decryption workflows. Only component with access to the content (ciphertext or plaintext).

> * *Virtru Identity Provider (IdP) (Keycloak)* - From Wikipedia:  "Keycloak is an open source software product to allow single sign-on with Identity and Access Management aimed at modern applications and services. As of March 2018 this JBoss community project is under the stewardship of Red Hat who use it as the upstream project for their RH-SSO product."  Keycloak is open source Apache License 2.0 code.

Reviewer: Can we talk about how it works with OIDC and IdPs in a generic term for the spec?

Reviewer 2: Yeah - see below for (a) suggestion

> * *Attribute Provider* (AP) - A web service that will receive requests from Virtru Protocol Mapper containing information about authenticated users and return our custom OIDC/OAuth claims in response.  It is the responsibility of Attribute Provider to transform 3rd party IdP claims/metadata to ABAC attributes.  It returns a Virtru [Claims Object](../schema/ClaimsObject.md).

Reviewer: Is the Virtru Protocol Mapper an open tool? Could we name it TDF Protocol Mapper or similar?

Reviewer 2: We actually probably don't need to talk about Virtru Protocol Mapper since it's a Keycloak-specific implementation/extension.

Perhaps instead of `VIrtru Identity Provider` we just say `OIDC Identity Provider with Custom Claims support`

Keycloak-plus-Protocol-Mapper is an example of 'OIDC Identity Provider with Custom Claims Support`

> * *Attribute Provider* (AP) - A web service that will receive requests from Virtru Protocol Mapper containing information about authenticated users and return our custom OIDC/OAuth claims in response.  It is the responsibility of Attribute Provider to transform 3rd party IdP claims/metadata to ABAC attributes.  It returns a Virtru [Claims Object](../schema/ClaimsObject.md).

Reviewer 2: ```suggestion
* *Attribute Provider* (AP) - A web service that will serve custom claims to an OIDC IdP with Custom Claims Support. The custom claims will include all custom claim data required by KAS to support TDF encryption and ABAC policy.  It is the responsibility of Attribute Provider to transform any 3rd party IdP claims/metadata to ABAC attributes.  It returns a TDF [Claims Object](../schema/ClaimsObject.md).
```

Suggestion on how to reword - feel free to take none/some of this.

> * *Key Access Service* (KAS) - Responsible for authorizing and granting TDF Clients access to rewrapped data key material. If authorized, TDF Clients can use this rewrapped data key to decrypt TDF ciphertext.

Reviewer 2: ```suggestion
* *Key Access Service* (KAS) - Responsible for authorizing and granting TDF Clients access to rewrapped data key material. If authorized, TDF Clients can use this rewrapped data key to decrypt TDF ciphertext. A valid OIDC token with [TDF Claims](../schema/ClaimsObject.md) must be used as a bearer token when communicating with KAS. An otherwise valid and trusted OIDC token without TDF claims will be rejected as a bearer token by KAS.
```

  * Any IdP software should work in this scenario provided it can execute some custom code on successful authentication that can:
    * Read the TDF Client public key from a custom HTTP header sent with the OIDC authentication request.
    * Construct and send an Attribute Provider web service request, including the public key in the payload.
    * Return the resulting Claims Object in the signed IdP JWT.
  * *Virtru Protocol Mapper* is our Keycloak-specific reference implementation of the above functionality.
* *Attribute Provider* (AP) - A web service that will receive requests from our OIDC IdP custom code (ex: Virtru Protocol Mapper) containing information about authenticated users and return our custom OIDC/OAuth claims in response.  It is the responsibility of Attribute Provider to transform 3rd party IdP claims/metadata to ABAC attributes.  It returns a TDF [Claims Object](../schema/ClaimsObject.md).

> * *Key Access Service* (KAS) - Responsible for authorizing and granting TDF Clients access to rewrapped data key material. If authorized, TDF Clients can use this rewrapped data key to decrypt TDF ciphertext.  A valid OIDC token containing [TDF Claims](../schema/ClaimsObject.md) must be used as a bearer token when communicating with KAS. An otherwise valid and trusted OIDC token without valid TDF Claims will be rejected as a bearer token by KAS.

Reviewer 3: Why will a valid OIDC token from a trusted IDP be rejected as a bearer token as opposed to just denying access to a policy due to not having the correct claims?

Reviewer 2: @biscoe916 it will do the latter, which is a more accurate way to describe what happens than `rejected as a bearer token, good point.

Reviewer 2: > Why will a valid OIDC token from a trusted IDP be rejected as a bearer token as opposed to just denying access to a policy due to not having the correct claims?

Good point. I might be wrong but I think there are two scenarios possible. 

- No TDF claims (i.e. no JSON key) in the bearer token KAS gets at all - can happen if you haven't configured Keycloak correctly, so it made sense to call it out somewhere as a misconfiguration.

- TDF claims exist in the bearer token KAS gets (i.e. there is a JSON key in the doc), but they aren't correct according to how policies have been set up for a given org - you don't have a misconfiguration in Keycloak, but your claims are still wrong.

"Rejected as a bearer token" was meant to indicate the first case - the second case would also involve KAS returning an error. Should both cases result in the same error, or should we distinguish?

## Workflow

The following sequence diagrams illustrate the workflow taken by the client to encrypt or decrypt TDF ciphertext. The canonical TDF architecture supports two modes of operation: _online mode_ and _offline mode_, which have distinct workflows as shown below.

_Online mode_ is the default mode, where the [wrapped data key](../schema/KeyAccessObject.md) and [authorization policy](../schema/PolicyObject.md) for TDF ciphertext is committed to KAS in-band as part of the `encrypt` operation. This means that the `encrypt` will succeed if and only if all resources are prepared to facilitate an immediate decrypt.

> _Offline mode_ requires that the TDF Client previously authenticated with OIDC to get a valid Claims Object. Clients running in this mode commit the [authorization policy](../schema/PolicyObject.md) out-of-band, or when decrypt is first performed. This significantly reduces latency at the cost of larger ciphertext (as the entire wrapped key and policy information must be included). 

Reviewer 3: The cipher text isn't larger. But there is a slightly larger manifest bundled with the TDF for the reasons mentioned here.

Also, I'm not sure that the client must have authenticated to get the claims object, unless something has changed that I'm not aware of.

Reviewer 3: To be clear, what I mean is that I'm not sure a claims object is required for offline encrypt. I didn't mean to suggest that one needn't be authenticated to get a claims object - something that is obviously false.

Reviewer 2: @biscoe916 At least for the preexisting JS SDK (`tdf3-js`), a network call to EAS to fetch a claims object is made unconditionally on every `encrypt` regardless of offline or online mode toggle - so that must have changed at some point.

The nanotdf and C++ SDKs support `setEntityObject` and will use the claims from a cached EO if present.

Currently the OIDC implementation **does** require the client to auth to get the claims object - in fact the client doesn't handle "claims objects" at all, it simply forwards the entire auth token it gets to KAS and doesn't have opinions about the content, or extract and parse any of it, by design - the bearer token and the client claims are one and the same.

The TDF3 clients however can simply be seeded with a JWT (previously obtained, containing claims) directly, which they can use as a KAS bearer token - this mimics the old `setEO`/offline claims model AFAICT.

Reviewer 2: ```
Long-lived bearer tokens can be made available for certain _offline mode_ use cases, for instance, sensors with intermittent or low-bandwidth connectivity can have an Entity Object flashed onto the device during manufacturing.
```

Obviously doing this with a JWT (or any clientside auth token/secret/credential/etc) is a terrible idea, so we might need to rethink this a bit if it's critical to support this flow with OIDC.

## Architecture

The canonical architecture contains four major components.

* *TDF Client* - Initiates and drives the TDF encryption and decryption workflows. Only component with access to the content (ciphertext or plaintext).

> * *OpenID Connect (OIDC) Identity Provider (IdP)* - This system could be any OIDC IdP software.  Virtru has chosen Keycloak as our reference implementation IdP.

Reviewer 2: The OpenID spec does not specify _how_ a compliant OIDC IdP obtains the custom claim data it inserts into OIDC tokens - and since getting the right claims into the token is essential for our stack (KAS) to work correctly, it makes sense to talk about how we actually accomplish this in a reference implementation.

> * *OpenID Connect (OIDC) Identity Provider (IdP)* - This system could be any OIDC IdP software.  Virtru has chosen Keycloak as our reference implementation IdP.

Author: The language choices we've made attempt to define the TDF spec and then our own reference implementation of that spec.  The goal would be to make the protocol less of a purely abstract concept by providing concrete examples and also to describe what a prospective OSS developer would receive when downloading our software and why each component is there.

  * From Wikipedia:  "Keycloak is an open source software product to allow single sign-on with Identity and Access Management aimed at modern applications and services. As of March 2018 this JBoss community project is under the stewardship of Red Hat who use it as the upstream project for their RH-SSO product."  Keycloak is open source Apache License 2.0 code.
  * Any IdP software should work in this scenario provided it can execute some custom code on successful authentication that can:
    * Read the TDF Client public key from a custom HTTP header sent with the OIDC authentication request.
    * Construct and send an Attribute Provider web service request, including the public key in the payload.
    * Return the resulting Claims Object in the signed IdP JWT.
  * *Virtru Protocol Mapper* is our Keycloak-specific reference implementation of the above functionality.

> * *Attribute Provider* (AP) - A web service that will receive requests from our OIDC IdP custom code (ex: Virtru Protocol Mapper) containing information about authenticated subjet and return our custom OIDC/OAuth claims in response.  It is the responsibility of Attribute Provider to transform 3rd party IdP claims/metadata to ABAC attributes.  It returns a TDF [Claims Object](../schema/ClaimsObject.md).

Reviewer 2: ```suggestion
* *Attribute Provider* (AP) - A web service that will receive requests from our OIDC IdP custom code (ex: Virtru Protocol Mapper) containing information about authenticated subjet and return our custom OIDC/OAuth claims in response.  It is the responsibility of Attribute Provider to transform incoming 3rd party IdP claims/metadata to outgoing ABAC policy attributes.  It returns a TDF [Claims Object](../schema/ClaimsObject.md).
```

## Workflow

The following sequence diagrams illustrate the workflow taken by the client to encrypt or decrypt TDF ciphertext. The canonical TDF architecture supports two modes of operation: _online mode_ and _offline mode_, which have distinct workflows as shown below.

_Online mode_ is the default mode, where the [wrapped data key](../schema/KeyAccessObject.md) and [authorization policy](../schema/PolicyObject.md) for TDF ciphertext is committed to KAS in-band as part of the `encrypt` operation. This means that the `encrypt` will succeed if and only if all resources are prepared to facilitate an immediate decrypt.

_Offline mode_ requires that the TDF Client previously obtained a

> valid, signed, long-lived OIDC JWT with embedded Claims Object via

Reviewer 4: ```suggestion
long-lived TDF JWT via
```

Reviewer 2: This change was applied (thanks!), so marking this as resolved.

_Online mode_ is the default mode, where the [wrapped data key](../schema/KeyAccessObject.md) and [authorization policy](../schema/PolicyObject.md) for TDF ciphertext is committed to KAS in-band as part of the `encrypt` operation. This means that the `encrypt` will succeed if and only if all resources are prepared to facilitate an immediate decrypt.

_Offline mode_ requires that the TDF Client previously authenticated with OIDC to get a valid Claims Object. Clients running in this mode commit the [authorization policy](../schema/PolicyObject.md) out-of-band, or when decrypt is first performed. This significantly reduces latency at the cost of larger ciphertext (as the entire wrapped key and policy information must be included). 

> ### Virtru IdP Direct Authentication, Encrypt

Reviewer 2: ```suggestion
### Hosted IdP Direct Authentication, Encrypt
```

> ![Virtru IdP Direct Authentication](../diagrams/OIDC_direct_auth.png)

Reviewer 2: ```suggestion
![Hosted IdP Direct Authentication](../diagrams/OIDC_direct_auth.png)
```

> ### Virtru IdP Direct Authentication, Decrypt

Reviewer 2: ```suggestion
### Hosted IdP Direct Authentication, Decrypt
```

> ![Virtru IdP Direct Authentication Decrypt](../diagrams/OIDC_direct_auth_decrypt.png)

Reviewer 2: ```suggestion
![Hosted IdP Direct Authentication Decrypt](../diagrams/OIDC_direct_auth_decrypt.png)
```

> ### Virtru IdP Token Exchange, Encrypt

Reviewer 2: ```suggestion
### Hosted IdP Token Exchange, Encrypt
```

> ![Virtru IdP Token Exchange](../diagrams/OIDC_token_exchange.png)

Reviewer 2: ```suggestion
![Hosted IdP Token Exchange](../diagrams/OIDC_token_exchange.png)
```

> #### Virtru IdP Brokered Authentication, Encrypt

Reviewer 2: ```suggestion
#### Hosted IdP Brokered Authentication, Encrypt
```

> ![Virtru IdP Brokered Authentication](../diagrams/OIDC_brokered_auth.png)

Reviewer 2: ```suggestion
![Hosted IdP Brokered Authentication](../diagrams/OIDC_brokered_auth.png)
```
